### PR TITLE
Version presence absence

### PR DESCRIPTION
**Pull request by thibaulttabarin**
Objective :
Compartmentalize the functionalities.
3 functionalities : 
+ Get presence absences of the traits
+ Extract measurement of traits 
+ Identified Landmarks

1- Command line tag to choose outputs
+ Add parser argument for documentation and modularity
+ By default output the presence absence dictionnary for Meghan Project
+ Optional positional args to output morphology measurement, landmarks and visualization

2- Add Comments

3- Improve Docker image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,19 +2,19 @@
 Extract morphological characteristics from image of fish trait segmentation.
 
 
-The goals of the tool is to extract measurments and landmarks of fish from the segmented fish iamge porduced by M. Maruf.
-It provides a framework for creating modularized tools by using classes and providing a way to visualize and test functionality using jupyter notebook. 
-We will release and containerize a working version for easy integration into the [BGNN_Snakemake](https://github.com/hdr-bgnn/BGNN_Snakemake).
-This tool can me made more generalizable but is a part of a bigger project, find the overview [Minnows Project](https://github.com/hdr-bgnn/minnowTraits).
+The goals of the tool is to extract presence-absence table, measurments and landmarks of fish from the segmented fish image porduced by M. Maruf.
```
Comment (johnbradley) [inline README.md]:
```suggestion
The goals of the tool is to extract presence-absence table, measurements and landmarks of fish from the segmented fish image produced by M. Maruf.
```

File: README.md [hunk 2]
```diff
@@ -2,19 +2,19 @@
...
+It provides a framework for creating modularized tools by using classes and jupyter notebook to visualize and test functionalities. 
+The tool is automatically containerized when a new release is published. It provide a validated version for easy integration into the [BGNN_Snakemake](https://github.com/hdr-bgnn/BGNN_Snakemake).
```
Comment (johnbradley) [inline README.md]:
```suggestion
The tool is automatically containerized when a new release is published. It provides a validated version for easy integration into the [BGNN_Snakemake](https://github.com/hdr-bgnn/BGNN_Snakemake).
```

File: README.md [hunk 3]
```diff
@@ -2,19 +2,19 @@
...
+This tool can me made more generalizable but it has been developped for the [Minnows Project](https://github.com/hdr-bgnn/minnowTraits).
```
Comment (johnbradley) [inline README.md]:
```suggestion
This tool can me made more generalizable but it has been developed for the [Minnows Project](https://github.com/hdr-bgnn/minnowTraits).
```

File: README.md [hunk 4]
```diff
@@ -2,19 +2,19 @@
...
+The segmented image input looks like image below, with traits color coded and identified by "blobs". The segmentation model uses [M. Maruf's segmentation code](https://github.com/hdr-bgnn/BGNN-trait-segmentation/blob/main/Segment_mini/scripts/segmentation_main.py), and is based on a Convolutional Neural Network (CNN; more specifically unet). You can find more information on the [BGNN-trait-segementation repository](https://github.com/hdr-bgnn/BGNN-trait-segmentation).  The output is 11 classes (traits) that are color coded. We are only using 9 of them, and are excluding alt_fin_ray and caudal_fin_ray.
```
Comment (johnbradley) [inline README.md]:
```suggestion
The segmented image input looks like image below, with traits color coded and identified by "blobs". The segmentation model uses [M. Maruf's segmentation code](https://github.com/hdr-bgnn/BGNN-trait-segmentation/blob/main/Segment_mini/scripts/segmentation_main.py), and is based on a Convolutional Neural Network (CNN; more specifically unet). You can find more information on the [BGNN-trait-segmentation repository](https://github.com/hdr-bgnn/BGNN-trait-segmentation).  The output is 11 classes (traits) that are color coded. We are only using 9 of them, and are excluding alt_fin_ray and caudal_fin_ray.
```

File: README.md [hunk 5]
```diff
@@ -91,30 +90,40 @@ We use the following descriptions and labels for landmarks and measurements. If
   angle  |  fish angle using landmarks  |  Fa_lm  |  angle of the tilt of the fish from horizontal (angle between the SL_lm and and the horizontal line of the image)  
   angle  |  fish angle using PCA  |  Fa_pca  |  angle of the tilt of the fish from horizontal (angle between the pca through the midline of the fish mask and the horizontal line of the image)  
   
-## 3- Trait measurement extraction
+## 3- Extract trait information : Classes
 
-We create a class to add more fexibility, which can be generalized to other projects. 
+We created classes to add more fexibility, which can be helpfull to generalize to other projects. [Trait_class](Scripts/Traits_class.py)
```
Comment (johnbradley) [inline README.md]:
```suggestion
We created classes to add more flexibility, which can be helpful to generalize to other projects. [Trait_class](Scripts/Traits_class.py)
```

File: README.md [hunk 6]
```diff
@@ -91,30 +90,40 @@ We use the following descriptions and labels for landmarks and measurements. If
...
-
-The output is a series of .json files with the fish number, traits, their measurements (in pixels), and the scale (in pixels/cm).
-
-We extract traits three ways:
-
-1) Using the landmarks
-2) Using the bbox
-3) Using the mask
++ Description : 
+  - Segmented_image class : This class creates an object "segmented_image" from [segmented_image.png](Scripts/test_images/INHS_FISH_000742_segmented.png). The object "segmented_image" importd the image.png and split into channels corresponding to the traits (trunk, dorsal_fin, etc.) in the form of dictionnary with key = trait  and value a mask. Then multiple functions will be applied to extract basic information on individual channel and on the whole fish.
```
Comment (johnbradley) [inline README.md]:
```suggestion
  - Segmented_image class : This class creates an object "segmented_image" from [segmented_image.png](Scripts/test_images/INHS_FISH_000742_segmented.png). The object "segmented_image" imported the image.png and split into channels corresponding to the traits (trunk, dorsal_fin, etc.) in the form of a dictionary with key = trait  and value a mask. Then multiple functions will be applied to extract basic information on individual channel and on the whole fish.
```

File: README.md [hunk 7]
```diff
@@ -91,30 +90,40 @@ We use the following descriptions and labels for landmarks and measurements. If
...
 
 #### 1) Using the landmarks
-These trait classes have the suffix "_lm" to denote the method of extraction. 
+These trait classes functions have the suffix "_lm" to denote the method of extraction. 
 
 The lengths (in pixels) are calculated as the distance between two landmarks (described in the "Definition" column of the trait description csvs).
 
 #### 2) Using the bbox (bounding box)
-These trait classes have the suffix "_bbox" to denote the method of extraction.
+These trait classes fucntiosn have the suffix "_bbox" to denote the method of extraction.
```
Comment (johnbradley) [inline README.md]:
```suggestion
These trait classes functions have the suffix "_bbox" to denote the method of extraction.
```

File: Scripts/Morphology_main.py [hunk 8]
```diff
@@ -6,49 +6,40 @@
...
-    f = open(metadata_file)
-    data = json.load(f)
-    metadata_fish = list(data.values())[0]['fish'][0]
-    fish_angle = None
-    major = []
-    length = []
     
-    if 'primary_axis' in metadata_fish  :
+    f = open(metadata_file)
+    metadata_dict = json.load(f)
```
Comment (johnbradley) [inline Scripts/Morphology_main.py]: Could we wrap this in a `with` clause to automatically close the file? I think this would leak a file handle.

**Review by johnbradley — CHANGES_REQUESTED** [review]
Code looks good to me. Please fix the typos in the README and close the leaking file handle.

File: Scripts/Morphology_main.py (hunk 8 above)
Comment (thibaulttabarin) [reply Scripts/Morphology_main.py]: good point. Sometime I do it sometime I don't... Thanks for the suggestion, make the code look better and more robust. It is done here
https://github.com/hdr-bgnn/Morphology-analysis/commit/0124795550ffdf5173e879338f24265ff3be6b64

**Review by johnbradley — APPROVED** [review]
LGTM

**Comment by thibaulttabarin** [comment]
In my view there is a last task that could be done, which would be to finish the jupyter notebook. Mainly cleaning, updating with the new changes and documentation.
This would be an enhancement... I will put it as issue.

File: Scripts/Morphology_main.py [hunk 9]
```diff
@@ -63,70 +54,95 @@ def default(self, obj):
...
-def main(input_file, metadata_file, output_measure, output_landmark, output_presence, 
-         output_lm_image=None):
+def argument_parser():
+    parser = argparse.ArgumentParser(description='Extract information from segmented fish image such as presence absence,\
+                                     landmarks, measures.')
+    parser.add_argument('input_image', help='Path of segmented fish image. Format PNG image file.')
+    parser.add_argument('output_presence', help='Path of output presence absence table. Format JSON file.')
+    
+    
+    parser.add_argument('--metadata', help='Path of input drexel_metadata_formatter. Format JSON metadata file.')
```
Comment (hlapp) [inline Scripts/Morphology_main.py]: It seems this refers to a metadata _file_, not a metadata _formatter_. Perhaps you mean to say that this is the path to the metadata file in  JSON format, structured as output by drexel_metadata_formatter?

File: Scripts/Morphology_main.py [hunk 10]
```diff
@@ -63,70 +54,95 @@ def default(self, obj):
...
+    Use Class Segmented_image, Measure_morphology to extract information 
+    from a segmented fish image (.png image).
+    
+    Input output are managed by argument_parser()
+    
+    Segmented_image creates an object that isolates different traits, preprocesses them and extract genral info (such as fish angle)
+    Measure_morphology inherits from Segmented_image and uses preprocessed traits information to measure morphology
+    characteristic and landmarks.
+    There are 4 mains output defined in argument_parser()
+    args.presence : {"dorsal_fin": {"number": 1, "percentage": 1.0}, "adipos_fin": {"number": 0, "percentage": 0}....}
```
Comment (hlapp) [inline Scripts/Morphology_main.py]: I think this is now named `output_presence`?

**Review by hlapp — COMMENTED** [review]
See inline comments. Aside from those, it seems the landmarking options and code are still here. Making sure that this is the intent?

File: Scripts/Morphology_main.py [hunk 11]
```diff
@@ -63,70 +54,95 @@ def default(self, obj):
...
+    number: number of blob per trait, percentage: area % of the bigger blob.
+    args.morphology : filename to save Morphology measurement from bbox and from lankmarks.
+    args.landmark : filename to save Coordinate of the landmark extracted.
+    args.lm_image : filename to save visualization of the landmarks
```
Comment (hlapp) [inline Scripts/Morphology_main.py:90]: Also, isn't the --metadata argument value missing here?

File: Scripts/Morphology_main.py (hunk 9 above)
Comment (thibaulttabarin) [reply Scripts/Morphology_main.py]: change as https://github.com/hdr-bgnn/Morphology-analysis/commit/6b2bdc1136c25945cd86cedfd91996ea1c319992

File: Scripts/Morphology_main.py (hunk 11 above)
Comment (thibaulttabarin) [reply Scripts/Morphology_main.py:90]: you are right. good catch!
https://github.com/hdr-bgnn/Morphology-analysis/commit/cd083c28908a2d0795b3c9f72dd9ad676c0db2d3

File: Scripts/Morphology_main.py (hunk 10 above)
Comment (thibaulttabarin) [reply Scripts/Morphology_main.py]: fixed 
https://github.com/hdr-bgnn/Morphology-analysis/commit/c718c28fbbac5edf31d85b437d31a877d84c6afc

**Comment by meghalithic** [comment]
@thibaulttabarin I left a section that says "XXXX" for you to fill in with how someone would use non-default functions

File: README.md [hunk 12]
```diff
@@ -91,70 +102,83 @@ We use the following descriptions and labels for landmarks and measurements. If
...
+    + landmark_image.png
+
+Usage:
 
-*Quick start*
-Create a segmented image object
 ```
-import Trait_class as tc 
-img_seg = tc.segmented_image("image_segmented.png")
+Morphology_main.py [-h] [--metadata METADATA] [--morphology MORPHOLOGY] [--landmark LANDMARK] [--lm_image LM_IMAGE] input_image output_presence
```
Comment (thibaulttabarin) [inline README.md]: @megbalk, I think that is what you want, this is the usage description on the function with 
+ The optional arguments : [-h] [--metadata METADATA] [--morphology MORPHOLOGY] [--landmark LANDMARK] [--lm_image LM_IMAGE] 
+ the positional arguments required : input_image output_presence

File: README.md [hunk 13]
```diff
@@ -30,24 +29,31 @@ When you export this image in python using pillow library (PIL.Image.open(file_n
...
+## 2- Presence Absence of Traits
+
+The approach that we use for checking presence and absence of traits is the following:
 
   1. Isolate indivual traits (e.g., isolate the dorsal_fin)
-  2. Remove small blobs and fill in gaps within each trait
-  3. Identify landmarks (defined in section 2)
-  4. Use landmarks and morphological tools (centroid, area, etc.) to extract the measurements (**external characters**)
+  2. Count number of blobs
+  3. Calculate percentage of large blob
```
Comment (thibaulttabarin) [inline README.md]: Calculate percentage area of the biggest blod over the total blobs of a particular trait

Comment (meghalithic) [reply README.md]: will make branch to fix readme

**Review by hlapp — APPROVED** [review]
I think this is good now?

**Pull request merged**